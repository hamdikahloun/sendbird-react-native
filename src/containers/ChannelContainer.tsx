import React, {useEffect, useRef} from 'react';
import {MessageInput, MessageList, ParticipantList} from '@/components';
import {useChannels} from '@/hooks/useChannels';
import {SafeAreaView, StyleSheet, TextInput} from 'react-native';
import {OpenChannel} from 'sendbird';
import {
  useUser,
  useParticipants,
  useOpenChannelMessageList,
  useStateChange,
  useChannelHandler,
  useChannel,
} from '@/hooks';

export const ChannelContainer = () => {
  const messageTextInputRef = useRef<TextInput>();
  const {user} = useUser();
  const {openChannel, setCurrentOpenChannel, sendMessage} = useChannel();
  useStateChange();
  useChannelHandler(openChannel);
  const {
    getOpenChannels,
    getGroupChannels,
    openChannels,
    openChannelsState,
    channelOpenListError,
  } = useChannels();

  const {
    channelOpenMessagesError,
    openChannelsMessagesState,
    openChannelMessages,
    getOpenChannelMessages,
  } = useOpenChannelMessageList();

  const {participants, getParticipantList} = useParticipants();
  const loading =
    openChannelsMessagesState === 'loading' || openChannelsState === 'loading';
  const error = channelOpenMessagesError || channelOpenListError;

  useEffect(() => {
    getOpenChannels();
    getGroupChannels();
  }, []);

  // For Testing purpose only
  useEffect(() => {
    init();
  }, [openChannels]);

  const init = async () => {
    if (openChannels.length && !openChannelMessages.length) {
      const firstChannel: OpenChannel = openChannels[0];
      await firstChannel.enter();
      setCurrentOpenChannel(firstChannel);
      getParticipantList(firstChannel);
      getOpenChannelMessages(firstChannel);
    }
  };

  // For Testing purpose only
  const onSendButtonPressed = (message: string) => {
    sendMessage(message);
    messageTextInputRef.current?.clear();
  };

  const onEndReached = () => {
    getOpenChannelMessages(openChannel);
  };

  return (
    <SafeAreaView style={styles.container}>
      <ParticipantList participants={participants} />
      <MessageList
        loading={loading}
        messages={openChannelMessages}
        error={error}
        connectedUserId={user.userId}
        onEndReached={onEndReached}
      />
      <MessageInput
        messageTextInputRef={messageTextInputRef}
        onSendButtonPressed={onSendButtonPressed}
      />
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#F5F5F5',
    paddingTop: 8,
  },
});
