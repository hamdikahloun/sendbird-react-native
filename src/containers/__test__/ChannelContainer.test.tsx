import React from 'react';
import {renderWithReduxProviders, waitFor} from '@/../testing';
import {ChannelContainer} from '../ChannelContainer';

describe('ChannelContainer', () => {
  it('should render correctly', async () => {
    const screen = renderWithReduxProviders(<ChannelContainer />).toJSON();
    await waitFor(() => expect(screen).toMatchSnapshot());
  });
});
