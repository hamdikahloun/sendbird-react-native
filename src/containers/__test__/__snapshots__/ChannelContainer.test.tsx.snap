// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChannelContainer should render correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#F5F5F5",
      "flex": 1,
      "paddingTop": 8,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#f3F3F3",
        "paddingBottom": 8,
        "paddingHorizontal": 16,
      }
    }
  >
    <RCTScrollView
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={true}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      showsHorizontalScrollIndicator={false}
      stickyHeaderIndices={Array []}
      viewabilityConfigCallbackPairs={Array []}
    >
      <View />
    </RCTScrollView>
  </View>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "backgroundColor": "#f3F3F3",
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      ListEmptyComponent={<ListEmptyComponent />}
      data={Array []}
      getItem={[Function]}
      getItemCount={[Function]}
      invertStickyHeaders={false}
      inverted={false}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onEndReached={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      showsVerticalScrollIndicator={false}
      stickyHeaderIndices={Array []}
      style={
        Object {
          "paddingHorizontal": 16,
        }
      }
      viewabilityConfigCallbackPairs={Array []}
    >
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <ActivityIndicator />
        </View>
      </View>
    </RCTScrollView>
  </RCTSafeAreaView>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "margin": 8,
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="Aa"
      style={
        Object {
          "backgroundColor": "#EFEFEF",
          "borderColor": "#F9F9F9",
          "borderRadius": 10,
          "borderWidth": 1,
          "flex": 1,
          "height": 50,
          "marginEnd": 8,
          "padding": 10,
        }
      }
      testID="message"
      value=""
    />
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "height": 48,
          "justifyContent": "center",
          "opacity": 1,
          "padding": 10,
          "width": 48,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../src/assets/images/send/send.png",
          }
        }
      />
    </View>
  </View>
</RCTSafeAreaView>
`;
